skimage/restoration/inpaint.py

```diff
@@ -1,79 +1,204 @@
+    """
+
+    n_channels = out.shape[-1]
+    radius = neigh_coef_full.shape[0] // 2
+
+    edge_mask = np.ones(mask.shape, dtype=bool)
+    edge_mask[(slice(radius, -radius),) * mask.ndim] = 0
+    boundary_mask = edge_mask * mask
+    center_mask = ~edge_mask * mask
+
+    boundary_pts = np.where(boundary_mask)
+    boundary_i = np.flatnonzero(boundary_mask)
+    center_i = np.flatnonzero(center_mask)
+    mask_i = np.concatenate((boundary_i, center_i))
+
+    center_pts = np.where(center_mask)
+    mask_pts = tuple(
+        [np.concatenate((b, c)) for b, c in zip(boundary_pts, center_pts)]
+    )
+
+    # Use convolution to predetermine the number of non-zero entries in the
+    # sparse system matrix.
+    structure = neigh_coef_full != 0
+    tmp = ndi.convolve(mask, structure, output=np.uint8, mode='constant')
+    nnz_matrix = tmp[mask].sum()
+
+    # Need to estimate the number of zeros for the right hand side vector.
+    # The computation below will slightly overestimate the true number of zeros
+    # due to edge effects (the kernel itself gets shrunk in size near the
+    # edges, but that isn't accounted for here). We can trim any excess entries
+    # later.
+    n_mask = np.count_nonzero(mask)
+    n_struct = np.count_nonzero(structure)
+    nnz_rhs_vector_max = n_mask - np.count_nonzero(tmp == n_struct)
+
+    # pre-allocate arrays storing sparse matrix indices and values
+    row_idx_known = np.empty(nnz_rhs_vector_max, dtype=np.intp)
+    data_known = np.zeros((nnz_rhs_vector_max, n_channels), dtype=out.dtype)
+    row_idx_unknown = np.empty(nnz_matrix, dtype=np.intp)
+    col_idx_unknown = np.empty(nnz_matrix, dtype=np.intp)
+    data_unknown = np.empty(nnz_matrix, dtype=out.dtype)
+
+    # cache the various small, non-square Laplacians used near the boundary
+    coef_cache = {}
+
+    # Iterate over masked points near the boundary
+    mask_flat = mask.reshape(-1)
+    out_flat = np.ascontiguousarray(out.reshape((-1, n_channels)))
+    idx_known = 0
+    idx_unknown = 0
+    mask_pt_n = -1
+    boundary_pts = np.stack(boundary_pts, axis=1)
+    for mask_pt_n, nd_idx in enumerate(boundary_pts):
         # Get bounded neighborhood of selected radius
-        b_lo, b_hi = _get_neighborhood(mask_pt_idx, 2, out.shape)
+        b_lo, b_hi = _get_neighborhood(nd_idx, radius, mask.shape)
 
-        # Create biharmonic coefficients ndarray
-        neigh_coef = np.zeros(b_hi - b_lo)
-        neigh_coef[tuple(mask_pt_idx - b_lo)] = 1
-        neigh_coef = laplace(laplace(neigh_coef))
+        # Create (truncated) biharmonic coefficients ndarray
+        coef_shape = tuple(b_hi - b_lo)
+        coef_center = tuple(nd_idx - b_lo)
+        neigh_coef, coef_idx, coefs = _get_neigh_coef(coef_shape,
+                                                      coef_center,
+                                                      coef_cache,
+                                                      dtype=out.dtype)
```

Comment:
What about removing the cache management from `_get_neigh_coef`?
I see here that the `neigh_coef` variable is not used in this function, so what about something like this:
```suggestion
        coef_idx, coefs = coef_cache.get((coef_shape, coef_center), (None, None)
        if coef_idx is None:
            _ , coef_idx, coefs = _get_neigh_coef(coef_shape,
                                                  coef_center
                                                  dtype=out.dtype)
            coef_cache[(coef_shape, coef_center)] = (coef_idx, coefs)
```
Doing so, the cache mechanism is explicit where it is necessary and the `_get_neigh_coef` is simplified... What do you think @grlee77?

Reply:
Yeah, that looks good.